src/modules/kvs/kvs_checkpoint.c

```diff
@@ -0,0 +1,230 @@
+{
+    if (kcp) {
+        double checkpoint_period = kcp->checkpoint_period;
+        if (checkpoint_period_parse (conf, errp, &checkpoint_period) < 0)
+            return -1;
+        kcp->checkpoint_period = checkpoint_period;
+    }
+    return 0;
+}
+
+int kvs_checkpoint_reload (kvs_checkpoint_t *kcp,
+                           const flux_conf_t *conf,
+                           flux_error_t *errp)
+{
+    if (kcp) {
+        double checkpoint_period = kcp->checkpoint_period;
+        if (checkpoint_period_parse (conf,
+                                     errp,
+                                     &checkpoint_period) < 0)
+            return -1;
+
+        if (checkpoint_period != kcp->checkpoint_period) {
+            kcp->checkpoint_period = checkpoint_period;
+            flux_watcher_stop (kcp->checkpoint_w);
+
+            if (kcp->root_primary
+                && kcp->checkpoint_period > 0.0) {
+                flux_timer_watcher_reset (kcp->checkpoint_w,
+                                          kcp->checkpoint_period,
+                                          kcp->checkpoint_period);
+                flux_watcher_start (kcp->checkpoint_w);
+            }
+        }
+    }
+    return 0;
+}
+
+
+static void checkpoint_cb (flux_reactor_t *r,
+                           flux_watcher_t *w,
+                           int revents,
+                           void *arg)
+{
+    kvs_checkpoint_t *kcp = arg;
+    char *name = NULL;
+    json_t *ops = NULL;
+
+    /* if no changes to root since last checkpoint-period, do
+     * nothing */
+    if (kcp->last_checkpoint_seq == kcp->root_primary->seq)
+        return;
+
+    if (asprintf (&name, "checkpoint-period.%u", kcp->root_primary->seq) < 0
+        || !(ops = json_array ())) {
+        errno = ENOMEM;
+        flux_log_error (kcp->h, "checkpoint-period setup failure");
+        goto error;
+    }
+
+    /* Set FLUX_KVS_SYNC, to perform the checkpoint.
+     *
+     * Set KVSTXN_API_FLAG_NO_PUBLISH, this is an internal KVS module
+     * transaction to checkpoint.  It has no operations so the KVS
+     * data will not change.  Therefore no setroot() needs to be
+     * called after this is done.
+     */
+    if (kvstxn_mgr_add_transaction (kcp->root_primary->ktm,
+                                    name,
+                                    ops,
+                                    FLUX_KVS_SYNC,
+                                    KVSTXN_API_FLAG_NO_PUBLISH) < 0) {
+        flux_log_error (kcp->h, "%s: kvstxn_mgr_add_transaction",
+                        __FUNCTION__);
+        goto error;
+    }
+
+    if (kcp->txn_cb)
+        kcp->txn_cb (kcp->root_primary, kcp->txn_cb_arg);
+
+    /* N.B. "last_checkpoint_seq" protects against unnecessary
+     * checkpointing when there is no activity in the primary KVS.
+     */
+    kcp->last_checkpoint_seq = kcp->root_primary->seq;
+
+ error:
+    free (name);
+    json_decref (ops);
```

Comment:
Clobbers errno

Reply:
does it matter in this case?  the function returns void.

Reply:
Oh sorry, I thought I was following a different call stack.  This is the timer watcher callback (duh).

Given that the function cannot return an error, maybe it would make sense to rename the goto label to "done".